File: Week_1/Day_3/3.py
#3.Container with Most Water
class Solution:
    def maxArea(self, height: list[int]) -> int:
        left=0
        right=len(height)-1
        max=0
        while left<right:
            width=right-left
            current=min(height[left],height[right])*width
            if current>max:
                max=current
                if height[left]>height[right]:
                    left+=1
                else:
                    right-=1
        return max

File: Week_3/Day_4/3.cpp
#include <vector>
#include <stack>
#include <algorithm>

class Solution {
public:
    int largestRectangleArea(std::vector<int>& heights) {
        std::stack<int> s;
        int maxArea = 0;
        int n = heights.size();

        for (int i = 0; i <= n; i++) {
            int currentHeight = (i == n) ? 0 : heights[i];
            while (!s.empty() && currentHeight < heights[s.top()]) {
                int height = heights[s.top()];
                s.pop();
                int width = s.empty() ? i : (i - s.top() - 1);

                maxArea = std::max(maxArea, height * width);
            }
            s.push(i);
        }

        return maxArea;
    }
};